spam/spam_tutorial.py

```diff
@@ -140,234 +145,406 @@
 # * *Third-party models*: using an pre-trained model (usually a model for a different task than the one at hand)
 # * *Distant supervision*: using external knowledge base
 # * *Crowdworker labels*: treating each crowdworker as a black-box function that assigns labels to subsets of the data
+
+# %% [markdown]
+# ### Recommended practice for LF development
+
+# %% [markdown]
+# Typical LF development cycles include multiple iterations of ideation, refining, evaluation, and debugging.
+# A typical cycle consists of the following steps:
+#
+# 1. Look at examples to generate ideas for LFs
+# 1. Write an initial version of an LF
+# 1. Estimate its performance by looking at examples in the training set or development set
+# 1. Refine and debug to improve coverage or accuracy as necessary
 #
-# The process of **developing LFs** is iterative and usually consists of:
-# * Writing a function
-# * Estimating its performance by looking at labeled examples in the training set or dev set
-# * Iterating to improve coverage or accuracy as necessary.
+# Labeling function development is a creative process.
+# However, there are a few principles that we highly recommend using when developing labeling functions.
 #
-# Balancing accuracy and coverage for specific labeling functions as well as the overall set of LFs developed is often a trade-off, and depending on the use case, users may tend to prefer one over the other.
+# * **Use the unlabeled training set for ideation whenever possible.** Reserve labeled examples for evaluation in order to avoid overfitting LFs or models.
+# * **Develop labeling functions in isolation.** Avoid using a "boosting" approach, trying to develop LFs specifically to interact with behavior of existing LFs. If you use the existing LFs to identify uncovered areas of the data, generalize the observed pattern rather than writing an LF to cover those exact examples.
+# * **Prioritize accuracy over coverage, except when accuracy is very high.** Very accurate LFs often result in higher performance models than ones with more coverage. However, coverage is often more difficult to obtain. So choose a 90% accuracy and 30% coverage version of an LF over a 95% accuracy and 10% coverage version. But choose an 80% accuracy LF and 20% coverage version of an LF over a 65% accuracy and 40% coverage version.
+# * **Try to cover as many data points as possible, but avoid overfitting.** In order to extract the most value out of your training set, regularly identify areas of the training set that haven't been covered yet and create new LFs that cover them. Don't obssess over covering the entire data set, as LFs written to cover outliers won't generalize.
+# * **Aim to have overlaps.** We'll train a _label model_ to denoise and combine the outputs of our LFs. The label model learns from overlaps and conflicts between LFs, so it's helpful if multiple LFs cover certain areas of the data set.
```

Comment:
I also think this is misleading and not really true- rather I'd say something like 'don't be afraid of overlaps'.

It is roughly true that the LabelModel will learn more when there are more overlaps... but we are concerned with overall performance of the full pipeline, not accuracy of the LabelModel parameter estimation or delta of LM > MV.  So not necessarily a good thing at all to explicitly aim for overlapping LFs.  Rather, just impress that we can handle them- don't need to go out of way to engineer non-overlapping LFs!